### PR TITLE
suggested tip amount based on cve number if a CVE is attached to an issue report

### PR DESCRIPTION
**Pull request by HanilJain**
# Feature Added
###  Issue #1928 

- **Added CVE Score** : If a CVE ID is attached to issue report it automatically fetches the CVE Score from this API Service https://services.nvd.nist.gov/rest/json/cves/2.0?cveId and saves to models 

- **functioning of Send tip feature** :  send tip feature was not working on /issue2/ page in this PR it was made functional 

- **Suggestion of tip** : On the basis of CVE Score, tip and severity is decided 
    -  cve_score >= 9.0  : critical : $50
    -  cve_score >= 7.0  : high : $20
    - cve_score >= 4.0 : medium : $10
    - cve_score >= 0.0 : low : $5 

 
Screenshot 2024-03-11 153934

### CODE REVIEW

File: website/templates/issue2.html [hunk 1]
```diff
@@ -536,6 +610,70 @@ <h2 class="mb-2 text-xl font-bold leading-tight text-gray-900 ">{{ issue.descrip
...
+        let suggestedTip;
+        if (severityLevel === "critical"){
+            suggestedTip = "$50";
+        } else if (severityLevel === "high") {
+            suggestedTip = "$20";
+        } else if (severityLevel === "medium") {
+            suggestedTip = "$10";
+        } else {
+            suggestedTip = "$5";
+        }
```
Comment (arkid15r) [inline website/templates/issue2.html:634]: It seems you can combine `severityLevel` and `suggestedTip` setting logic:
```suggestion
        let severityLevel;
        let suggestedTip;

        if (cve_score >= 9.0) {
            severityLevel = "critical";
            suggestedTip = "$50";
        } else if (cve_score >= 7.0) {
            severityLevel = "high";
            suggestedTip = "$20";
        } else if (cve_score >= 4.0) {
            severityLevel = "medium";
            suggestedTip = "$10";
        } else {
            severityLevel = "low";
            suggestedTip = "$5";
        }
```

Comment (HanilJain) [reply website/templates/issue2.html]: @arkid15r ohh I missed that , ThanksYou

File: website/models.py [hunk 2]
```diff
@@ -312,6 +313,22 @@ def remove_user(self):
     def get_absolute_url(self):
         return "/issue/" + str(self.id)
 
+    def get_cve_score(self):
+        if self.cve_id:
```
Comment (arkid15r) [inline website/models.py:317]: Consider checking `if self.cve_id is None` first:
```suggestion
        if self.cve_id is None:
            return None
```
to save an indent below

File: website/models.py [hunk 3]
```diff
@@ -312,6 +313,22 @@ def remove_user(self):
...
+                response = requests.get(url).json()
+                results = response["resultsPerPage"]
+                if results != 0:
+                    metrics = response["vulnerabilities"][0]["cve"]["metrics"]
+                    if metrics:
+                        cvss_metric_v = next(iter(metrics))
+                        score = metrics[cvss_metric_v][0]["cvssData"]["baseScore"]
+                        self.score = score
+                    return score
+            except Exception as e:
```
Comment (arkid15r) [inline website/models.py:329]: Could you narrow down the expected exceptions set? It's not a good practice to catch broad exception.

File: website/models.py [hunk 4]
```diff
@@ -312,6 +313,22 @@ def remove_user(self):
...
+            try:
+                url = "https://services.nvd.nist.gov/rest/json/cves/2.0?cveId=%s" % (self.cve_id)
+                response = requests.get(url).json()
+                results = response["resultsPerPage"]
+                if results != 0:
+                    metrics = response["vulnerabilities"][0]["cve"]["metrics"]
+                    if metrics:
+                        cvss_metric_v = next(iter(metrics))
+                        score = metrics[cvss_metric_v][0]["cvssData"]["baseScore"]
+                        self.score = score
```
Comment (arkid15r) [inline website/models.py:327]: It should either return the `score` or set it, not both. Based on the function name I'd go with return option.

**Comment by HanilJain** [comment]
@arkid15r  I have made all the suggested changes

File: website/models.py [hunk 5]
```diff
@@ -312,6 +313,26 @@ def remove_user(self):
     def get_absolute_url(self):
         return "/issue/" + str(self.id)
 
+    def save_cve_score(self):
```
Comment (arkid15r) [inline website/models.py:316]: I liked the old name better. The current name doesn't reflect what the method does (technically it doesn't save the score but rather gets it via API and puts into `self.score`).

File: website/models.py [hunk 6]
```diff
@@ -312,6 +313,26 @@ def remove_user(self):
...
+                if metrics:
+                    cvss_metric_v = next(iter(metrics))
+                    score = metrics[cvss_metric_v][0]["cvssData"]["baseScore"]
+                    self.score = score
+        except requests.exceptions.HTTPError as e:
+            print(e)
+            return None
+        except requests.exceptions.ReadTimeout as e:
+            print(e)
+            return None
```
Comment (arkid15r) [inline website/models.py:334]: Could you group these exceptions:
```suggestion
        except (requests.exceptions.HTTPError, requests.exceptions.ReadTimeout) as e:
            print(e)
            return None
```

File: website/models.py (hunk 5 above)
Comment (HanilJain) [reply website/models.py:316]: ohkay !!

**Comment by HanilJain** [comment]
@arkid15r thanks for the suggestion, please check

File: website/migrations/0087_issue_cve_score.py
Comment (arkid15r) [inline website/migrations/0087_issue_cve_score.py:1]: I've just realized we already have `score`. Do you use both fields to save the same CVE base score value?

Comment (HanilJain) [reply website/migrations/0087_issue_cve_score.py:1]: @arkid15r  damnn it was typo it meant to be 
`self.cve_score = score `
thanks for mentioning it I have made the changes

File: website/models.py [hunk 7]
```diff
@@ -312,6 +313,23 @@ def remove_user(self):
...
+        try:
+            url = "https://services.nvd.nist.gov/rest/json/cves/2.0?cveId=%s" % (self.cve_id)
+            response = requests.get(url).json()
+            results = response["resultsPerPage"]
+            if results != 0:
+                metrics = response["vulnerabilities"][0]["cve"]["metrics"]
+                if metrics:
+                    cvss_metric_v = next(iter(metrics))
+                    score = metrics[cvss_metric_v][0]["cvssData"]["baseScore"]
+                    self.cve_score = score
```
Comment (arkid15r) [inline website/models.py:328]: My original concern was regarding redundant setting `self.cve_score` here as you also doing it in the views.py L884. Consider just returning it as a result of `Issue::get_cve_score()`. Other option would be rename it to `set_cve_score()`.

**Review by arkid15r — COMMENTED** [review]
Yes, looks better!

**Review by DonnieBLT — CHANGES_REQUESTED** [review]
one small tweak from me

File: website/models.py (hunk 7 above)
Comment (HanilJain) [reply website/models.py:328]: Done 👍

File: website/templates/issue2.html [hunk 8]
```diff
@@ -536,6 +610,65 @@ <h2 class="mb-2 text-xl font-bold leading-tight text-gray-900 ">{{ issue.descrip
...
+        { name: 'Bitcoin', address: btc },
+        { name: 'Ethereum', address: eth },
+        { name: 'Bitcoin Cash', address: bch }
+        ];
+        let currentIndex = 0;
+
+        const updateSlide = () => {
+            let { name, address } = cryptoOptions[currentIndex];
+            if (address == '' || address == 'None') {
+                address = "qr5yccf7j4dpjekyz3vpawgaarl352n7yv5d5mtzzc";
```
[the remark this answers is not in the document]
Comment (HanilJain) [reply website/templates/issue2.html:642]: image

File: website/models.py [hunk 9]
```diff
@@ -312,6 +313,23 @@ def remove_user(self):
...
+        try:
+            url = "https://services.nvd.nist.gov/rest/json/cves/2.0?cveId=%s" % (self.cve_id)
+            response = requests.get(url).json()
+            results = response["resultsPerPage"]
+            if results != 0:
+                metrics = response["vulnerabilities"][0]["cve"]["metrics"]
+                if metrics:
+                    cvss_metric_v = next(iter(metrics))
+                    score = metrics[cvss_metric_v][0]["cvssData"]["baseScore"]
+                    return score
```
Comment (arkid15r) [inline website/models.py:328]: Optional: another optimization you could use is 
```suggestion
                    return metrics[cvss_metric_v][0]["cvssData"]["baseScore"]
```
right away as you're using `score` variable just once.

**Review by arkid15r — COMMENTED** [review]
Awesome, thank you @HanilJain!

Looks good!

File: website/models.py (hunk 9 above)
Comment (HanilJain) [reply website/models.py:328]: ohh yess thanks

**Review by DonnieBLT — APPROVED** [review]

**Pull request merged**